### PR TITLE
Manual addition of `structured_pattern` constraints on class mentioning slots, plus...

### PR DESCRIPTION
**Pull request by turbomam**
- retire `local/nmdc-schema-v7.8.0.yaml`
- removal of some old commented stuff in the schema files
- `Activity` and `WorkflowExecutionActivity` become `abstract`

### CODE REVIEW

**Comment by turbomam** [comment]
There are still many appearances of 'xx' or 'generic:' in the valid examples data, especially for `has_raw_value` and `has_output`

**Comment by turbomam** [comment]
@aclum 

with the addition of the mentioning slot pattern constraints, `nmdc_schema/nmdc_schema_accepting_legacy_ids.yaml` doesn't really accept mentioned legacy ids anymore.  That's really apparent in `local/mongo_as_nmdc_database_validation.log` generated by `make-rdf`

**Comment by turbomam** [comment]
`nmdc_schema/nmdc_schema_accepting_legacy_ids.yaml` is tolerant of legacy `id`s in any position now

File: src/data/valid/Database-extraction_set-exhaustive.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@ extraction_set:
     description: DNA extraction of NEON sample WREF_072-O-20190618-COMP using SOP BMI_dnaExtractionSOP_v7
     #enforce analytical sample typecode?
     has_input:
-      - generic:xxx
+      - nmdc:procsm-99-123456
     #cardinality of outputs should be 1-many, may want to fractionate samples or keep supernatant
     has_output:
       - generic:xxx
```
Comment (aclum) [inline src/data/valid/Database-extraction_set-exhaustive.yaml:13]: If we've done things correctly generic:xxx should not be valid for has_output for this class. It should only allow an instance of Class ProcessedSample.

File: src/data/valid/Database-extraction_set-minimal.yaml [hunk 2]
```diff
@@ -3,7 +3,7 @@ extraction_set:
   - id: nmdc:extrp-99-abcdef
     name: DNA extraction of NEON sample WREF_072-O-20190618-COMP
     has_input:
-      - generic:xxx
+      - nmdc:procsm-99-123456
     #cardinality of outputs should be 1-many, may want to fractionate samples or keep supernatant
     has_output:
       - generic:yyy
```
Comment (aclum) [inline src/data/valid/Database-extraction_set-minimal.yaml:9]: Same comments as on src/data/valid/Database-extraction_set-exhaustive.yaml, this example data will need to be updated once the range of has_output is fixed.

File: src/data/valid/Database-library-prep-exhausive.yaml [hunk 3]
```diff
@@ -5,7 +5,7 @@ library_preparation_set:
     name: DNA library creation of NEON sample TREE_001-O-20170707-COMP-DNA1
     description: DNA extraction of NEON sample TREE_001-O-20170707-COMP-DNA1 using SOP BMI_metagenomicsSequencingSOP_v2
     has_input:
-      - generic:xxx
+      - nmdc:bsm-99-987654
     has_output:
       - generic:xxx
```
Comment (aclum) [inline src/data/valid/Database-library-prep-exhausive.yaml:10]: Same comments here Class LIbraryConstruction, class range for slot_has output needs to be updated to [ProcessedSample](https://microbiomedata.github.io/berkeley-schema-fy24/ProcessedSample/) and example data needs to be updated.

File: src/data/valid/Database-metagenome-assembly.yaml [hunk 4]
```diff
@@ -1,7 +1,8 @@
 metagenome_assembly_set:
   - id: nmdc:wfmgas-99-B7Vogx.1
     name: Metagenome assembly 1472_51277
-    was_informed_by: nmdc:omprc-11-547rwq94
+    was_informed_by: nmdc:omprc-11-547rwq94 # MAM 2024-05-22
+    #    was_informed_by: nmdc:act-11-547rwq94.1 # should this allow an omics processing as the informer? what about act? that's abstract isn't it?
```
Comment (aclum) [inline src/data/valid/Database-metagenome-assembly.yaml:5]: They way this slot is being used in practice is to tie WorkflowExecutionActivity subclass instances to OmicsProcessing & this is what the data portal uses The instances themselves are linked by has_input and has_output. We shouldn't make any further changes to this now b/c WorkflowChain in Berkeley schema addresses this.

File: src/data/valid/Database-nucleic-extraction.yaml [hunk 5]
```diff
@@ -13,7 +13,7 @@ extraction_set:
     #      the range of has_input for this class, as a slot_usage
     # this still doesn't solve referential integrity
     has_input:
-      - generic:xxx
+      - nmdc:procsm-99-123456
 
     #cardinality of outputs should be 1-many, may want to fractionate samples or keep supernatant
     has_output:
```
Comment (aclum) [inline src/data/valid/Database-nucleic-extraction.yaml:19]: Same as PlannedProcess subclasses, range for has_output needs to be updated to PlannedProcess and example data needs to be updated.

File: src/data/valid/Database-pooling_set-exhaustive.yaml [hunk 6]
```diff
@@ -5,8 +5,8 @@ pooling_set:
       - generic:xxx
     description: xxx
     has_input:
-      - generic:xxx
-      - generic:yyy
+      - nmdc:bsm-99-123456
+      - nmdc:bsm-99-987654
     has_output:
       - generic:xxx
```
Comment (aclum) [inline src/data/valid/Database-pooling_set-exhaustive.yaml:11]: Pooling has_output range needs to be updated to ProcessedSample and example data updated.

File: src/data/valid/Extraction-proteomics.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 id: "nmdc:extrp-38-r2pk"
 has_input:
-  - "nmdc:fasp-37"
+  - nmdc:procsm-99-abcdef
 has_output:
   - "nmdc:ome-39"
```
Comment (aclum) [inline src/data/valid/Extraction-proteomics.yaml:5]: Same comment here, this has_output value should not be valid under any circumstances b/c there is no blade and will become invalid once you update the range of has_output to PlannedProcess.

File: src/data/valid/Pooling-exhaustive.yaml [hunk 8]
```diff
@@ -4,8 +4,8 @@ alternative_identifiers:
   - generic:xxx
 description: xxx
 has_input:
-  - generic:xxx
-  - generic:yyy
+  - nmdc:bsm-99-123456
+  - nmdc:bsm-99-987654
 has_output:
   - generic:xxx
```
Comment (aclum) [inline src/data/valid/Pooling-exhaustive.yaml:10]: Will need to fix this example data.

File: src/schema/nmdc.yaml [hunk 9]
```diff
@@ -225,6 +230,9 @@ classes:
         any_of:
           - range: Biosample
           - range: ProcessedSample
+        structured_pattern:
+          syntax: "{id_nmdc_prefix}:(bsm|procsm)-{id_shoulder}-{id_blade}"
```
Comment (aclum) [inline src/schema/nmdc.yaml:234]: Missing closing $

File: src/data/valid/DataObject-exhaustive.yaml [hunk 10]
```diff
@@ -10,4 +10,4 @@ md5_checksum: "22afa3d49b73eaec2e9787a6b88fbdc3"
 name: Crispr Terms
 type: "nmdc:DataObject"
 url: http://example.com
-was_generated_by: "nmdc:invalid_id"
+was_generated_by: "nmdc:act-99-123456.1"
```
Comment (aclum) [inline src/data/valid/DataObject-exhaustive.yaml:13]: Update to WorkflowExecutionActivity subclass instance. Activity is abstract so won't mint an identifier.

File: src/schema/nmdc.yaml [hunk 11]
```diff
@@ -362,6 +376,10 @@ classes:
         structured_pattern:
           syntax: "{id_nmdc_prefix}:dobj-{id_shoulder}-{id_blade}$"
           interpolated: true
+      was_generated_by:
+        structured_pattern:
+          syntax: "{id_nmdc_prefix}:(act|wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfnom|wfrbt|wfrqc|wf|omprc)-{id_shoulder}-{id_blade}{id_version}$"
```
Comment (aclum) [inline src/schema/nmdc.yaml:381]: remove act, Activity is an abstract class now so won't mint IDs.

File: src/schema/prov.yaml [hunk 12]
```diff
@@ -40,6 +41,11 @@ classes:
         structured_pattern:
           syntax: "{id_nmdc_prefix}:act-{id_shoulder}-{id_blade}{id_version}$"
           interpolated: true
+      was_informed_by:
+        structured_pattern:
+          #          syntax: "{id_nmdc_prefix}:(act|wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfnom|wfrbt|wfrqc|wf)-{id_shoulder}-{id_blade}{id_version}$"
```
Comment (aclum) [inline src/schema/prov.yaml:46]: remove 'act', Activity is abstract now.

**Review by aclum — CHANGES_REQUESTED** [review]
We need to tighten the constraints on the range for has_output on PlannedProcess subclasses to ProcessedSample and remove references to Activity as an identifier b/c that is abstract now. See individual comments for details.

File: src/data/valid/Extraction-metabolomics.yaml [hunk 13]
```diff
@@ -1,6 +1,6 @@
 id: "nmdc:extrp-71-r2pk"
 has_input:
-  - "nmdc:ome-6"
+  - nmdc:procsm-99-abcdef
 has_output:
   - "nmdc:ome-8"
```
[the remark this answers is not in the document]
Comment (turbomam) [reply src/data/valid/Extraction-metabolomics.yaml:5]: isn't 'abcdef' the blade? 'nmdc' is the prefix, 'procsm' is the typecode, '99' is the minting authority, and  'abcdef' is the blade, right?

File: src/schema/nmdc.yaml [hunk 14]
```diff
@@ -362,6 +398,10 @@ classes:
         structured_pattern:
           syntax: "{id_nmdc_prefix}:dobj-{id_shoulder}-{id_blade}$"
           interpolated: true
+      was_generated_by:
+        structured_pattern:
+          syntax: "{id_nmdc_prefix}:(wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfnom|wfrbt|wfrqc|wf|omprc)-{id_shoulder}-{id_blade}{id_version}$"
```
Comment (aclum) [inline src/schema/nmdc.yaml:403]: 'wf' should be removed, WorkflowExeuctionActivity is abstract. id_version should be allowed on workflow activity subclasses but not OmicsProcessing records so I think this should be re-written like this
syntax: "{id_nmdc_prefix}:(wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfnom|wfrbt|wfrqc)-{id_shoulder}-{id_blade}{id_version}$ | {id_nmdc_prefix}:omprc-{id_shoulder}-{id_blade}$"

File: src/schema/prov.yaml [hunk 15]
```diff
@@ -76,6 +77,9 @@ slots:
       - range: OmicsProcessing
     mappings:
       - prov:wasGeneratedBy
+    structured_pattern:
```
Comment (aclum) [inline src/schema/prov.yaml:80]: Same comments as on slot was_generated_by on Class DataObject. 'wf' should be removed, WorkflowExeuctionActivity is abstract. id_version should be allowed on workflow activity subclasses but not OmicsProcessing records so I think this should be re-written like this
syntax: "{id_nmdc_prefix}:(wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfnom|wfrbt|wfrqc)-{id_shoulder}-{id_blade}{id_version}$ | {id_nmdc_prefix}:omprc-{id_shoulder}-{id_blade}$"
Can we just define this globally on the slot and then remove the local definition on Class DataObject?

**Pull request merged**